closes #23504: download focus wallpapers at runtime

MatthewTighe opened the pull request:
### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Amejia481 reviewed (commented):
Did a first pass it looks good in general, just added some suggestions

Amejia481 commented on app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt:
```diff
@@ -0,0 +1,83 @@
+                    input.copyTo(localFile.outputStream())
+                }
+            }.onFailure {
+                logger.error(it.message ?: "download failed: no throwable message", it)
```
This will be log as info in crash systems

MatthewTighe replied:
Updated!

Amejia481 commented on app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperManager.kt:
```diff
@@ -46,13 +60,31 @@ class WallpaperManager(
+                loadWallpaperFromDisk(context, newWallpaper)
+            }
+            if (bitmap == null) {
+                logger.error("Could not load wallpaper bitmap. Resetting to default")
```
Similar to https://github.com/mozilla-mobile/fenix/pull/23505#discussion_r796798226

MatthewTighe replied:
Added a crash report!

gabrielluong commented on app/src/main/java/org/mozilla/fenix/settings/wallpaper/WallpaperSettings.kt:
```diff
@@ -225,12 +228,12 @@ private fun WallpaperThumbnailItem(
-            val contentDescription = stringResource(
-                R.string.wallpapers_item_name_content_description, wallpaper.name
-            )
+        val contentDescription = stringResource(
```
Perhaps inline the contentDescription now that it's not conditional

MatthewTighe replied:
Updated!

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/wallpapers/Wallpaper.kt:
```diff
@@ -24,6 +27,57 @@ data class Wallpaper(
+ * Types defining whether a [Wallpaper] is delivered through a remote source or is included locally
+ * in the APK.
+ */
+sealed class WallpaperOrigin {
+    object Local : WallpaperOrigin()
+    object Remote : WallpaperOrigin()
+}
```
Oh yeah, I like that! We could even designate them as `Local` or `Remote` as part of the type hierarchy. Something like:

```
sealed class Wallpaper(...) {
    sealed class Local(...) : Wallpaper(...){
        data class FirefoxWallpaper(...) : Local(...)
        ...
    }
}
```

But what if we refactor it to use the `State` pattern eventually? I'm not immediately finding examples of any `State` objects that have `sealed class` properties. A `sealed class` has the same issue that an `enum class` does, in that we could define an `object` child with global mutable state, something like 

```
sealed class Wallpaper(...) {
    sealed class Local(...) : Wallpaper(...){
        object Default : Local {
            var resourceId: Int // changes based on app light or dark theme
        ...
    }
}
```

It seemed like we wanted to avoid using anything but nested `data class`es for state, when this came up in my first attempt at the refactor in #23113. Are `sealed class`es okay, but not `enum class`es? What do you think @gabrielluong ?

gabrielluong replied:
My conclusion back then was that they are both fined. The main gotcha is what I presented in our slack discussion. Ideally, we just don't have too much functionality within the class.

MatthewTighe replied:
Okay cool. I'll update this in the follow-up PR that moves the wallpapers from `asset`s to `drawable`s, as it'll be easier when not relying on the dynamic data contained in `wallpapers.json`.

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/FenixApplication.kt:
```diff
@@ -781,4 +782,11 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
+
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun downloadWallpapers() {
+        GlobalScope.launch {
```
I like to switch the dispatcher closest to the code that actually needs to be on the different dispatcher. That way, calling code doesn't need to introspect the async code to discover it needs to change dispatchers. This is a [suggested practice](https://developer.android.com/kotlin/coroutines/coroutines-adv#main-safety), and is something that can be observed in [Room](https://medium.com/androiddevelopers/room-coroutines-422b786dc4c5) and in Retrofit I believe. 

That is, it's safe to call `suspend` functions defined in Room and Retrofit from any `Dispatcher`, because the libraries ensure they switch to `IO` before performing async tasks. I like to mirror that behavior.

Amejia481 replied:
We have some previous discussion about using the `Dispatchers.Default` when was strictly needed  you could read more https://github.com/mozilla-mobile/fenix/issues/19574 @mcomella and @csadilek could provide more pointers as they were more involved on the ticket.

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/wallpapers/Wallpaper.kt:
```diff
@@ -24,6 +27,57 @@ data class Wallpaper(
+/**
+ * Get the expected local path on disk for a wallpaper if orientation and app theme are known.
+ */
+fun Wallpaper.getLocalPath(orientation: String, theme: String): String =
+    "$orientation/$theme/${themeCollection.directoryName()}/$name.png"
+
+private fun Context.isLandscape(): Boolean {
+    return resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE
+}
+
+private fun Context.isDark(): Boolean {
+    return resources.configuration.uiMode and
+        Configuration.UI_MODE_NIGHT_MASK == Configuration.UI_MODE_NIGHT_YES
 }
```
I could move these extensions into that file if that makes more organizational sense, but `Wallpaper.getLocal` path is also used by `WallpaperDownloader`. `WallpaperDownloader` is a child of `WallpaperManager`, and so should not receive it as a dependency.

I could also just copy the functions into both, if that makes more sense. What do you think?

Amejia481 replied:
We could also have a `Extensions.kt` file under the `wallpapers` package similar to what we do here

https://github.com/mozilla-mobile/android-components/blob/main/components/feature/addons/src/main/java/mozilla/components/feature/addons/ui/Extensions.kt

The would be covered in follow up, let's leave it as it's for now.

MatthewTighe commented:
Added a commit addressing feedback, will squash them before landing. @Amejia481 and I decided that we'll land a separate PR for switching the local wallpapers from `assets` to `drawables`, and that will include the discussed changes to the `Wallpaper` type.

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperManager.kt:
```diff
@@ -46,13 +60,31 @@ class WallpaperManager(
+                loadWallpaperFromAssets(newWallpaper, context)
+            } else {
+                loadWallpaperFromDisk(context, newWallpaper)
+            }
```
Oh good call, encapsulated this into a single function

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt:
```diff
@@ -0,0 +1,83 @@
+                client.fetch(request)
+            }.onSuccess {
+                if (!it.isSuccess) {
+                    logger.error("download response failure code: ${it.status}")
```
Capitalized!

MatthewTighe commented on app/src/main/java/org/mozilla/fenix/wallpapers/Wallpaper.kt:
```diff
@@ -24,6 +27,57 @@ data class Wallpaper(
+ * Types defining whether a [Wallpaper] is delivered through a remote source or is included locally
+ * in the APK.
+ */
+sealed class WallpaperOrigin {
```
Switch this and `WallpaperThemeCollection` to both be enums

Amejia481 commented on app/src/main/java/org/mozilla/fenix/wallpapers/Wallpaper.kt:
```diff
@@ -23,7 +26,51 @@ data class Wallpaper(
+/**
+ * Get the parent directory name for a remote wallpaper asset.
+ */
+fun WallpaperThemeCollection.directoryName(): String = when (this) {
+    WallpaperThemeCollection.NONE,
+    WallpaperThemeCollection.FIREFOX -> ""
+    WallpaperThemeCollection.FOCUS -> "focus"
+}
```
nit: This could be a property on the `WallpaperThemeCollection` :) 
We could do that in follow up.

MatthewTighe replied:
A better name might actually be what we want here. This function returns the remote directory name, so I don't think it's necessarily an applicable property for all theme collections

Amejia481 commented on app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperDownloader.kt:
```diff
@@ -0,0 +1,85 @@
+                }
+            }.onFailure {
+                logger.error(it.message ?: "Download failed: no throwable message included.", it)
+                context.components.analytics.crashReporter.submitCaughtException(it)
```
nit: For future testing purpose we could the reference of `crashReporter` to  `WallpaperDownloader`,  we could do this in a follow up.

Amejia481 commented on app/src/main/java/org/mozilla/fenix/FenixApplication.kt:
```diff
@@ -127,6 +127,7 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
 
         setupInMainProcessOnly()
 
+        downloadWallpapers()
```
I almost missed this one, this should be under a feature validation `FeatureFlags.showWallpapers` + `FeatureFlags.isThemedWallpapersFeatureEnabled`.

Amejia481 replied:
Maybe we could add `FeatureFlags.isThemedWallpapersFeatureEnabled` later

MatthewTighe replied:
🙏 good catch, TYVM

Amejia481 reviewed (approved):
Tested locally and it's working like a charm!
